src/Elastic.Extensions.Logging.Common/LogEventBuilderExtensions.cs

```diff
@@ -0,0 +1,212 @@
+using System.Collections;
+using System.Text;
+using Elastic.CommonSchema;
+using Microsoft.Extensions.Logging;
+
+namespace Elastic.Extensions.Logging.Common;
+
+/// <summary> </summary>
+public static class LogEventBuilderExtensions
+{
+	/// <summary> </summary>
+	public static void AddScopeValues(this LogEvent logEvent, IExternalScopeProvider? scopeProvider, ILogEventCreationOptions options)
+	{
+		void AddScopeValue<TLocalState>(TLocalState scope, LogEvent log)
+		{
+			if (scope is null) return;
+
+			log.Labels ??= new Labels();
+			log.Scopes ??= new List<string>();
+
+
+			var scopeValues = (scope as IEnumerable<KeyValuePair<string, object>>)?.ToList();
+			var scopeName = scopeValues != null && scopeValues.Any(kv => kv.Key == "{OriginalFormat}")
+				? scope.ToString()
+				: FormatValue(scope, options, 0, scope.GetType().Name);
+			log.Scopes.Add(scopeName);
+
+			if (scopeValues == null) return;
+
+			foreach (var kvp in scopeValues)
+				AssignStateOrScopeLabels(logEvent, kvp, options);
+		}
+
+		scopeProvider?.ForEachScope((o, @event) => AddScopeValue(o, @event), logEvent);
+	}
+
+	/// <summary> </summary>
+	public static void AddStateValues<TState>(this LogEvent logEvent, TState state, ILogEventCreationOptions options)
+	{
+		if (state is not IEnumerable<KeyValuePair<string, object>> stateValues) return;
+
+		foreach (var kvp in stateValues)
+			AssignStateOrScopeLabels(logEvent, kvp, options);
+	}
+
+
+	private static void AssignStateOrScopeLabels(LogEvent logEvent, KeyValuePair<string, object> kvp, ILogEventCreationOptions options)
+	{
+		if (kvp.Key == "{OriginalFormat}")
+		{
+			// we explicitly want this to override, preferring OriginalFormat from current state over scope
+			logEvent.MessageTemplate = kvp.Value.ToString();
+			return;
+		}
+		var value = FormatValue(kvp.Value, options);
+		if (!AssignKnownHttpKeys(logEvent, kvp.Key, value))
+			logEvent.AssignField(kvp.Key, value);
+	}
+
+	private static bool AssignKnownHttpKeys(LogEvent logEvent, string key, object value)
+	{
+		switch (key)
+		{
+			case "RequestId" when value is string requestId:
+				logEvent.Http ??= new Http();
+				logEvent.Http.RequestId = requestId;
+				return true;
+			case "RequestPath" when value is string path:
+				logEvent.Url ??= new Url();
+				logEvent.Url.Path = path;
+				return true;
+			// ReSharper disable once UnusedVariable
+			case "Protocol" when value is string protocol:
+				// TODO protocol
```

Comment:
Was this missed or for follow up?

Reply:
Ahh this was moved pasted from elsewhere into the `Common` lib. 

I think we have no place to put it currently so to the TODO is not actionable